test/results_test.dart

```diff
@@ -662,4 +662,42 @@ Future<void> main([List<String>? args]) async {
       if (count > 1) fail('Should only receive one event');
     }
   });
+
+  test('RealmResults.first throws on empty', () async {
+    final config = Configuration.local([Dog.schema, Person.schema]);
+    final realm = getRealm(config);
+
+    final empty = Iterable<int>.empty();
+    expect(() => empty.first, throwsA(isA<StateError>()));
+    expect(() => realm.all<Dog>().first, throwsA(isA<RealmException>())); // TODO: Should this be StateError?
+  });
+
+  test('RealmResults.last throws on empty', () async {
+    final config = Configuration.local([Dog.schema, Person.schema]);
+    final realm = getRealm(config);
+
+    final empty = Iterable<int>.empty();
+    expect(() => empty.last, throwsA(isA<StateError>()));
+    expect(() => realm.all<Dog>().last, throwsA(isA<RealmException>())); // TODO: Should this be StateError?
+  });
+
+  test('RealmResult.single throws on empty', () {
+    final config = Configuration.local([Dog.schema, Person.schema]);
+    final realm = getRealm(config);
+
+    final empty = Iterable<int>.empty();
+    expect(() => empty.single, throwsA(isA<StateError>()));
+    expect(() => realm.all<Dog>().single, throwsA(isA<StateError>()));
```

Comment:
I think this should throw `RealmException`

Reply:
throws `RealmStateError`